### PR TITLE
feat: add module for feature flagging

### PR DESCRIPTION
**Pull request by odinr**
## Why
__What kind of change does this PR introduce?__
adds a module for using feature flagging

closes: #1210

- [x] create module for feature flagging
- [x] add plugin for session storage
- [x] add plugin for CGI (url parameters)
- [x] add README

### Check off the following:
- [x] Confirm that I checked changes to branch which I am merging into.
  - _I have validated included files_
  - _My code does not generate new linting warnings_
  - _My PR is not a duplicate, [check existing pr`s](https://github.com/equinor/fusion-framework/pulls)_
- [x] Confirm that the I have completed the [self-review checklist](https://github.com/equinor/fusion-framework/blob/main/contributing/self-review.md).

- [x] Confirm that my changes meet our [code of conduct](https://github.com/equinor/fusion-framework/blob/main/CODE_OF_CONDUCT.md).

#### Conditional
- [x] Confirm project selected and category, actual, iteration are set
- [x] Confirm Milestone selected _(if any)_

### CODE REVIEW

**Comment by odinr** [comment]
no change-log atm, since we don`t want to release this package yet.

**Review by eikeland — APPROVED** [review]

File: packages/cli/src/bin/dev-portal/PersonSideSheet/index.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    const Component = useMemo(() => {
+        let Comp;
+        switch (currentSheet) {
+            // case 'allocations':
+            //     Comp = AllocationSheetContent;
+            //     break;
+            // case 'roles':
+            //     Comp = RoleSheetContent;
+            //     break;
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/index.tsx:32]: remove

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureSheetContent.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import { FeatureTogglerApp } from './FeatureTogglerApp';
+import { FeatureTogglerPortal } from './FeatureTogglerPortal';
+
+import { Divider, Icon, Button, Tabs } from '@equinor/eds-core-react';
+import { arrow_back, category } from '@equinor/eds-icons';
+Icon.add({ arrow_back, category });
+
+import type { SheetContentProps } from './types';
+
+export const FeatureSheetContent = ({ navigate }: SheetContentProps) => {
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureSheetContent.tsx:11]: short comment on what this component is ment todo

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerApp.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import { useCurrentAppFeatureFlags } from '@equinor/fusion-framework-react/feature-flag';
+
+import { Typography, Switch } from '@equinor/eds-core-react';
+
+import { Styled } from './Styled';
+
+export const FeatureTogglerApp = () => {
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerApp.tsx:7]: short comment

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerPortal.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { useFrameworkFeatureFlags } from '@equinor/fusion-framework-react/feature-flag';
+
+import { Typography, Switch } from '@equinor/eds-core-react';
+
+import { Styled } from './Styled';
+
+export const FeatureTogglerPortal = () => {
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerPortal.tsx:7]: short comment

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/LandingSheetContent.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                        <Icon name="work_outline"></Icon>
+                        My Allocations
+                    </Button>
+                </BtnListItem>
+                <BtnListItem>
+                    <Button variant="ghost" onClick={() => navigate('roles')}>
+                        <Icon name="security"></Icon>
+                        My Roles
+                    </Button>
+                </BtnListItem> */}
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/LandingSheetContent.tsx:32]: remove

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/LandingSheetContent.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+const BtnList = styled.ul`
+    list-style: none;
+    padding-left: 0;
+`;
+const BtnListItem = styled.li`
+    margin: 1em 0;
+`;
+
+export const LandingSheetContent = ({ azureId, navigate }: SheetContentProps) => {
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/LandingSheetContent.tsx:17]: short comment

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/types.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// import type { ApiPerson_v4 } from '@equinor/fusion-framework-module-services/people/api-models.v4';
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/types.ts:1]: remove

File: packages/cli/src/bin/dev-portal/config.ts [hunk 8]
```diff
@@ -38,6 +39,21 @@ export const configure = async (config: FrameworkConfigurator) => {
         });
     });
 
+    enableFeatureFlagging(config, (builder) => {
```
Comment (odinr) [inline packages/cli/src/bin/dev-portal/config.ts:42]: add short comment that this is only for demo, needs to support by conf in future

File: packages/cli/src/bin/dev-portal/PersonSideSheet/index.tsx (hunk 1 above)
Comment (eikeland) [reply packages/cli/src/bin/dev-portal/PersonSideSheet/index.tsx:32]: done

File: packages/modules/feature-flag/src/FeatureFlagModule.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+export const module: FeatureFlagModule = {
+    name,
+    configure: () => new FeatureFlagConfigurator(),
+    initialize: async (init) => {
+        const config = await init.config.createConfigAsync(init);
+        return new FeatureFlagProvider(config);
+    },
+};
+
+export const enableFeatureFlagging = (
```
Comment (odinr) [inline packages/modules/feature-flag/src/FeatureFlagModule.ts:18]: add doc what this does, with an example

File: packages/modules/feature-flag/src/FeatureFlagProvider.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    findFeature,
+} from './FeatureFlagProvider.selectors';
+
+import type { FeatureFlagConfig } from './types';
+import { IFeatureFlag } from './FeatureFlag';
+import { normalizeFlags } from './utils/normalize-flags';
+
+export type CompareFeature = (a: IFeatureFlag | undefined, b: IFeatureFlag | undefined) => boolean;
+
+export interface IFeatureFlagProvider {
```
Comment (odinr) [inline packages/modules/feature-flag/src/FeatureFlagProvider.ts:18]: add doc to all methods

File: packages/modules/feature-flag/src/plugins/api/create-client.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import type { IHttpClient } from '@equinor/fusion-framework-module-http';
+import type { ApiResponseSelector, IApiPluginClient } from './types';
+
+export const createApiPluginClient = (args: {
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/api/create-client.ts:4]: missing doc

Comment (odinr) [reply packages/modules/feature-flag/src/plugins/api/create-client.ts:4]: and example

File: packages/modules/feature-flag/src/plugins/api/create-plugin.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
+import type { ApiFeatureFlagPlugin, ApiResponseSelector } from './types';
+import { IFeatureFlag, FeatureFlag } from '../../FeatureFlag';
+import { FeatureFlagPluginConfigCallback } from '../../types';
+import { createApiPluginClient } from './create-client';
+import { ApiPlugin } from './plugin';
+
+const defaultSelector: ApiResponseSelector = async (response: Response) => {
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/api/create-plugin.ts:7]: short comment

File: packages/modules/feature-flag/src/plugins/api/create-plugin.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+            value: flag.value,
+            title: flag.title,
+            description: flag.description,
+            source: flag.source,
+            readonly: true,
+        });
+    });
+};
+
+export const createApiPlugin = (args: {
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/api/create-plugin.ts:22]: short comment

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureSheetContent.tsx (hunk 2 above)
Comment (eikeland) [reply packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureSheetContent.tsx:11]: Done

File: packages/modules/feature-flag/src/plugins/api/types.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import type { ObservableInput } from 'rxjs';
+import type { IFeatureFlag } from '../../FeatureFlag';
+
+export type { FeatureFlagPlugin as ApiFeatureFlagPlugin } from '../../types';
+
+export interface IApiPluginClient {
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/api/types.ts:6]: missing docs

File: packages/modules/feature-flag/src/plugins/cgi/create-client.ts [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import { IFeatureFlag } from '../../FeatureFlag';
+import createStorage, { IStorageAdapter } from '../../utils/storage';
+import { ICgiPluginClient } from './types';
+import { name as namespace } from '../../FeatureFlagModule';
+
+export type PluginOptions = {
+    storage?: IStorageAdapter<IFeatureFlag>;
+};
+
+export const createCgiPluginClient = (
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/cgi/create-client.ts:10]: missing docs

File: packages/modules/feature-flag/src/plugins/cgi/create-plugin.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
+import { INavigationProvider } from '@equinor/fusion-framework-module-navigation';
+import { FeatureFlagPluginConfigCallback } from '../../types';
+import type { CgiFeatureFlagPlugin } from './types';
+import CgiPlugin from './plugin';
+import { createCgiPluginClient } from './create-client';
+import { IFeatureFlag } from '../../FeatureFlag';
+
+export const createCgiPlugin = (
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/cgi/create-plugin.ts:8]: missing docs

File: packages/modules/feature-flag/src/plugins/cgi/plugin.ts [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+            }),
+        );
+    }
+
+    onFeatureToggle(event: { flags: Array<IFeatureFlag> }) {
+        const { flags } = event;
+        this.#cgiClient.storeFeatureFlags(flags);
+    }
+
+    connect(args: { provider: IFeatureFlagProvider }) {
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/cgi/plugin.ts:64]: missing docs, comment code what happens on connect

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerApp.tsx (hunk 3 above)
Comment (eikeland) [reply packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerApp.tsx:7]: Done

File: packages/modules/feature-flag/src/plugins/cgi/plugin.ts [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+                                acc.push({ key, enabled });
+                            }
+                            return acc;
+                        },
+                        [] as Array<{ key: string; enabled: boolean }>,
+                    ),
+                );
+            }),
+            takeUntil(teardown$),
+        );
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/cgi/plugin.ts:90]: add doc

File: packages/modules/feature-flag/src/plugins/cgi/types.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    value: string | null;
+    /** path of hit */
+    path: Path;
+};
+
+export type AssertFeatureFlag = (options: CgiAssertionOptions) => boolean;
+
+export type { FeatureFlagPlugin as CgiFeatureFlagPlugin } from '../../types';
+
+export interface ICgiPluginClient {
```
Comment (odinr) [inline packages/modules/feature-flag/src/plugins/cgi/types.ts:19]: missing docs

File: packages/modules/feature-flag/src/utils/storage/StorageAdapter.ts [hunk 20]
```diff
@@ -0,0 +1,82 @@
+const ID_SEPARATOR = '::';
+const NAME_SEPARATOR = '_';
+
+const extractId = (key: string): { basename: string; id: string } => {
+    const [basename, id] = key.split(ID_SEPARATOR);
+    return { basename, id };
+};
+
+export interface IStorageAdapter<TType = unknown> extends Iterable<TType> {
```
Comment (odinr) [inline packages/modules/feature-flag/src/utils/storage/StorageAdapter.ts:9]: missing docs

File: packages/modules/feature-flag/src/utils/storage/index.ts [hunk 21]
```diff
@@ -0,0 +1,24 @@
+import { StorageAdapter, type IStorageAdapter } from './StorageAdapter';
+
+import { LocalStorageAdapter } from './LocalStorageAdapter';
+import { SessionStorageAdapter } from './SessionStorageAdapter';
+
+export { type IStorageAdapter, StorageAdapter, LocalStorageAdapter, SessionStorageAdapter };
+
+export type StorageType = 'local' | 'session';
+
+export const createStorage = <TType = unknown>(
```
Comment (odinr) [inline packages/modules/feature-flag/src/utils/storage/index.ts:10]: docs

Comment (odinr) [reply packages/modules/feature-flag/src/utils/storage/index.ts:10]: with example

File: packages/react/app/package.json [hunk 22]
```diff
@@ -4,34 +4,14 @@
...
-            "types": "./dist/types/navigation/index.d.ts"
-        }
+        ".": "./dist/esm/index.js",
+        "./bookmark": "./dist/esm/bookmark/index.js",
+        "./context": "./dist/esm/context/index.js",
+        "./feature-flag": "./dist/esm/feature-flag/index.js",
+        "./framework": "./dist/esm/framework/index.js",
+        "./http": "./dist/esm/http/index.js",
+        "./msal": "./dist/esm/msal/index.js",
+        "./navigation": "./dist/esm/navigation/index.js"
```
Comment (odinr) [inline packages/react/app/package.json:14]: why was this changed, this is not uniformed with normal export

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerPortal.tsx (hunk 4 above)
Comment (eikeland) [reply packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/FeatureTogglerPortal.tsx:7]: done

File: packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/LandingSheetContent.tsx (hunk 5 above)
Comment (eikeland) [reply packages/cli/src/bin/dev-portal/PersonSideSheet/sheets/LandingSheetContent.tsx:32]: Done

File: packages/cli/src/bin/dev-portal/config.ts (hunk 8 above)
Comment (eikeland) [reply packages/cli/src/bin/dev-portal/config.ts:42]: Done

File: packages/modules/feature-flag/src/FeatureFlagModule.ts (hunk 9 above)
Comment (eikeland) [reply packages/modules/feature-flag/src/FeatureFlagModule.ts:18]: done

**Comment by asbjornhaland** [comment]
@odinr Can you have a check and resolve the conversations that you are happy with?

**Review by asbjornhaland — COMMENTED** [review]
LGTM

**Comment by asbjornhaland** [comment]
@odinr @eikeland Are we ready to squash/merge this now, or?

**Review by eikeland — APPROVED** [review]

**Review by asbjornhaland — APPROVED** [review]
